core/src/test/java/org/apache/iceberg/TestSequenceNumberForV2Table.java

```diff
@@ -0,0 +1,328 @@
+            manifestEntry(ManifestEntry.Status.ADDED, null, FILE_D)))
+        .commit();
+
+    manifestFile = table.currentSnapshot().manifests().stream()
+        .filter(manifest -> manifest.snapshotId() == table.currentSnapshot().snapshotId())
+        .collect(Collectors.toList()).get(0);
+    validateManifestEntries(manifestFile, 4, files(FILE_A, FILE_B, FILE_C, FILE_D), seqs(1, 2, 3, 4));
+    validateDataFiles(files(FILE_A, FILE_B, FILE_C, FILE_D), seqs(1, 2, 3, 4));
+  }
+
+  @Test
+  public void testRewrite() {
+    table.newFastAppend().appendFile(FILE_A).commit();
+    table.newFastAppend().appendFile(FILE_B).commit();
+    Assert.assertEquals(2, table.currentSnapshot().sequenceNumber());
+
+    table.rewriteManifests().clusterBy(file -> "").commit();
+    Assert.assertEquals("Snapshot sequence number should be 3",
+        3, table.currentSnapshot().sequenceNumber());
+
+    ManifestFile newManifest = table.currentSnapshot().manifests().stream()
+        .filter(manifest -> manifest.snapshotId() == table.currentSnapshot().snapshotId())
+        .collect(Collectors.toList()).get(0);
+
+    validateManifestEntries(newManifest, 3, files(FILE_A, FILE_B), seqs(1, 2));
+  }
+
+  @Test
+  public void testCommitConflict() {
+    AppendFiles appendA = table.newFastAppend();
+    appendA.appendFile(FILE_A).apply();
+
+    table.updateProperties()
+        .set(TableProperties.COMMIT_NUM_RETRIES, "0")
+        .commit();
+
+    table.ops().failCommits(1);
+
+    AssertHelpers.assertThrows("Should reject commit",
+        CommitFailedException.class, "Injected failure",
+        () -> table.newFastAppend().appendFile(FILE_B).commit());
+
+    table.updateProperties()
+        .set(TableProperties.COMMIT_NUM_RETRIES, "5")
+        .commit();
+
+    appendA.commit();
+
+    Assert.assertEquals(1, table.currentSnapshot().sequenceNumber());
+
+    AppendFiles appendFiles = table.newFastAppend().appendFile(FILE_C);
+    appendFiles.apply();
+    table.newFastAppend().appendFile(FILE_D).commit();
+    appendFiles.commit();
+
+    validateDataFiles(files(FILE_A, FILE_D, FILE_C), seqs(1, 2, 3));
+  }
+
+  @Test
+  public void testRollBack() {
+    table.newFastAppend().appendFile(FILE_A).commit();
+    long snapshotId = table.currentSnapshot().snapshotId();
+
+    table.newFastAppend().appendFile(FILE_B).commit();
+    Assert.assertEquals("Snapshot sequence number should match expected",
+        2, table.currentSnapshot().sequenceNumber());
+
+    table.manageSnapshots().rollbackTo(snapshotId).commit();
+    Assert.assertEquals("Snapshot sequence number should match expected",
+        1, table.currentSnapshot().sequenceNumber());
+
+    Assert.assertEquals("Table last sequence number should be 2",
+        2, table.operations().current().lastSequenceNumber());
+
+    table.newFastAppend().appendFile(FILE_C).commit();
+    Assert.assertEquals("Snapshot sequence number should match expected",
+        3, table.currentSnapshot().sequenceNumber());
```

Comment:
Overall, this is a good test. But you do need to add validations for the rest of the metadata.